### PR TITLE
Simplify generic job reconciler factory

### LINKED ISSUE #1593: Flaky E2E test for Pod replacement
**What happened**:

Sometimes the Workload is never declared as Finished.

```
End To End Suite: kindest/node:v1.29.0: [It] Pod groups when Single CQ Failed Pod can be replaced in group expand_less	50s
{Timed out after 45.000s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/util/util.go:300 with:
it's finished failed [FAILED] Timed out after 45.000s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/util/util.go:300 with:
it's finished
In [It] at: /home/prow/go/src/sigs.k8s.io/kueue/test/e2e/singlecluster/pod_test.go:224 @ 01/16/24 20:19:48.709
}
```

Looking at the kube-scheduler logs, the Pod called `excess` got scheduled, which shouldn't have happened. This either means that some other Pod failed or was identified as an excess by Kueue.

**What you expected to happen**:

The workload to finish successfully

**How to reproduce it (as minimally and precisely as possible)**:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/1592/pull-kueue-test-e2e-main-1-29/1747351345448357888

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Simplifies the factory for genericJobReconciler.
The Pod integration no longer uses the genericJobReconciler, as it behaves significantly different.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

I will use this new PodReconciler to fix #1593

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/jobs/pod/event_handlers.go [hunk 1]
```diff
@@ -73,23 +73,23 @@ func (h *podEventHandler) queueReconcileForPod(ctx context.Context, object clien
 	q.Add(reconcileRequestForPod(p))
 }
 
-type parentWorkloadHandler struct{}
+type workloadHandler struct{}
```
Comment (alculquicondor) [inline pkg/controller/jobs/pod/event_handlers.go:76]: In theory, this is not necessary, but the name was not accurate.

File: pkg/controller/jobframework/reconciler.go [hunk 2]
```diff
@@ -946,42 +946,36 @@ func GetPodSetsInfoFromWorkload(wl *kueue.Workload) []podset.PodSetInfo {
 
 }
 
-// NewGenericReconciler creates a new reconciler factory for a concrete GenericJob type.
+type ReconcilerSetup func(*builder.Builder, client.Client) *builder.Builder
```
Comment (tenzen-y) [inline pkg/controller/jobframework/reconciler.go:949]: Oh, this is a cool implementation :)

File: pkg/controller/jobs/pod/pod_controller.go [hunk 3]
```diff
@@ -86,17 +85,26 @@ func init() {
...
-		return &Pod{}
-	},
-	func(c client.Client) (handler.EventHandler, string) {
-		return &podEventHandler{}, "v1_pod"
-	},
-	func(c client.Client) handler.EventHandler {
-		return &parentWorkloadHandler{}
-	},
-)
+type Reconciler struct {
```
Comment (tenzen-y) [inline pkg/controller/jobs/pod/pod_controller.go:88]:
```suggestion
var _ jobframework.JobReconcilerInterface = (*Reconciler)(nil)

type Reconciler struct {
```

This `Reconciler` implements `jobframework.JobReconcilerInterface`, right?

Comment (alculquicondor) [reply pkg/controller/jobs/pod/pod_controller.go:88]: No need for this. If Reconciler failed to implement the interface, then `NewReconciler` below wouldn't compile

Comment (tenzen-y) [reply pkg/controller/jobs/pod/pod_controller.go:88]: It makes sense.
/lgtm
/approve

**Pull request merged**